File: src/main/java/gameoflife/app/GameOfLifeUI.java
package gameoflife.app;

import gameoflife.algorithm.GameOfLife;
import gameoflife.helper.IOHelper;
import gameoflife.helper.SeedHelper;

import javax.swing.JComponent;
import javax.swing.JFrame;
import javax.swing.WindowConstants;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Insets;
import java.awt.KeyEventPostProcessor;
import java.awt.KeyboardFocusManager;
import java.awt.Toolkit;
import java.awt.event.KeyEvent;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class GameOfLifeUI extends JComponent implements KeyEventPostProcessor {
    private static final int WAIT_TIME = 100;
    private static final int MAX_CELL_SIZE = 100;
    private static final int MIN_CELL_SIZE = 6;
    private static final String OPT_STEP = "-s";
    private static final String OPT_WAIT = "-w";
    private static final Logger LOG = Logger.getLogger(GameOfLifeUI.class.getName());

    private final SeedHelper seedHelper = new SeedHelper();
    private GameOfLife gameOfLife;
    private final JFrame window = new JFrame();
    private int cellSize = MAX_CELL_SIZE;
    private boolean continueFlag = true;
    private int evolveToggle = 1;
    private boolean automaton = true;
    private String path;
    private Boundary boundary;
    private int iteration;
    private int waitTime;

    public GameOfLifeUI(String[] params) {
        Optional.of(params)
                .filter(p -> p.length > 0)
                .map(this::setup)
                .orElseThrow(() -> new RuntimeException("Missing seeds"));
    }

    private GameOfLifeUI setup(String[] params) {
        path = params[0];
        gameOfLife = buildGameOfLife(IOHelper.loadSeeds(path));
        automaton = isAutomaton(params);
        waitTime = getWaitTime(params);
        return this;
    }

    private GameOfLife buildGameOfLife(String[] seeds) {
        boundary = seedHelper.getBoundaryFromHeader(seeds);
        return new GameOfLife(seedHelper.seedToMap(seeds));
    }

    private int getWaitTime(String[] params) {
        return Stream.of(params)
                .filter(param -> param.startsWith(OPT_WAIT))
                .map(param -> Integer.parseInt(param.substring(OPT_WAIT.length())))
                .findFirst()
                .orElse(WAIT_TIME);
    }

    public GameOfLifeUI setWaitTime(int waitTime) {
        this.waitTime = waitTime;
        return this;
    }

    private boolean isAutomaton(String[] params) {
        return !Arrays.asList(params).contains(OPT_STEP);
    }

    public boolean isContinueFlag() {
        return continueFlag;
    }

    public GameOfLife getGameOfLife() {
        return gameOfLife;
    }

    void run() {
        setupKeyboardListener();
        setupFrame();

        while (isContinueFlag()) {
            repaint();
            waitAWhile();
            if (isContinueToEvolve()) {
                evolve();
            }
        }

        close();
    }

    private boolean isContinueToEvolve() {
        return automaton || evolveToggle == 0;
    }

    private void close() {
        window.setVisible(false);
        window.dispose();
    }

    private void setupKeyboardListener() {
        KeyboardFocusManager.getCurrentKeyboardFocusManager()
                .addKeyEventPostProcessor(this);
    }

    private void setupFrame() {
        setCellSize(calculateCellSize());

        final int width = calculatePanelSize(boundary.getX());
        final int height = calculatePanelSize(boundary.getY());

        setSize(width, height);
        setFocusable(true);
        window.setTitle(getTitle(path));
        window.setResizable(false);
        window.setVisible(true);
        window.setBounds(getHorizontalPosition(width), getVerticalPosition(height),
                         getFrameWidth(width), getFrameHeight(height));
        window.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        window.getContentPane().add(this);
    }

    private int calculateCellSize() {
        final Dimension screenSize = getScreenSize();
        return Math.max(Math.min(Math.min(calculateCellSize(screenSize.height, boundary.getY()),
                                          calculateCellSize(screenSize.width, boundary.getX())),
                                 MAX_CELL_SIZE),
                        MIN_CELL_SIZE);
    }

    private int calculateCellSize(int screenSize, double numberOfCells) {
        return (int)(screenSize * 3 / 4 / numberOfCells);
    }

    private int calculatePanelSize(double position) {
        return (int)(position * cellSize);
    }

    private int getHorizontalPosition(int panelWidth) {
        return calculatePosition(getScreenSize().width, getFrameWidth(panelWidth));
    }

    private int getVerticalPosition(int panelHeight) {
        return calculatePosition(getScreenSize().height, getFrameHeight(panelHeight));
    }

    private int calculatePosition(int screenSize, int frameSize) {
        return (screenSize - frameSize) / 2;
    }

    private Dimension getScreenSize() {
        return Toolkit.getDefaultToolkit().getScreenSize();
    }

    private int getFrameWidth(int panelWidth) {
        return panelWidth + calculateInsertsValue(insets -> insets.left + insets.right);
    }

    private int getFrameHeight(int panelHeight) {
        return panelHeight + calculateInsertsValue(insets -> insets.top + insets.bottom);
    }

    private int calculateInsertsValue(Function<Insets, Integer> getter) {
        return getter.apply(window.getInsets());
    }

    private void evolve() {
        gameOfLife = gameOfLife.tick();
        evolveToggle++;
        window.setTitle(String.format("%s - #%d", path, iteration));
        iteration++;
    }

    private int getCellPosition(int index) {
        return index * cellSize;
    }
    private int getFillPosition(int index) {
        return getCellPosition(index) + 1;
    }
    private int getFillSize() {
        return cellSize - 2;
    }

    private Color getColor(int x, int y) {
        return gameOfLife.isActive(x, y) ? getForeground() : getBackground();
    }

    private final Function<Graphics, Function<Integer, Consumer<Integer>>> fillCell = graphics -> y -> x -> {
        graphics.setColor(getColor(x, y));
        graphics.fillRect(getFillPosition(x), getFillPosition(y), getFillSize(), getFillSize());
    };

    private final Function<Graphics, Function<Integer, Consumer<Integer>>> drawBorder = graphics -> y -> x -> {
        graphics.setColor(getForeground());
        graphics.drawRect(getCellPosition(x), getCellPosition(y), cellSize, cellSize);
    };

    private final List<Function<Graphics, Function<Integer, Consumer<Integer>>>> LAMBDAS = Arrays.asList(fillCell, drawBorder);

    private void paintRow(Consumer<Integer> actor) {
        IntStream.range(0, boundary.getX()).forEach(actor::accept);
    }
    private void paintRows(Function<Integer, Consumer<Integer>> paint) {
        IntStream.range(0, boundary.getY()).forEach(y -> paintRow(paint.apply(y)));
    }

    @Override
    public void paint(Graphics graphics) {
        LAMBDAS.forEach(lambda -> paintRows(lambda.apply(graphics)));
    }

    private void waitAWhile() {
        if (waitTime > 0) {
            try {
                synchronized (this) {
                    wait(waitTime);
                }
            } catch (InterruptedException ex) {
                LOG.log(Level.SEVERE, ex.getMessage(), ex);
            }
        }
    }

    private void setCellSize(int value) {
        this.cellSize = value;
    }

    @Override
    public boolean postProcessKeyEvent(KeyEvent e) {
        boolean result = false;
        if (e.getKeyCode() == KeyEvent.VK_ESCAPE) {
            continueFlag = false;
        } else if (e.getKeyCode() == KeyEvent.VK_SPACE) {
            automaton = e.isControlDown();
            evolveToggle = evolveToggle < 2 ? evolveToggle + 1 : 0;
            e.consume();
            result = true;
        }
        return result;
    }

    private String getTitle(String path) {
        return String.format("Seed: %s", path);
    }

    public static void main(String[] params) {
        new GameOfLifeUI(params).run();
    }
}

File: src/main/kotlin/gameoflife/app/GameOfLifeUI.kt
package gameoflife.app

import gameoflife.algorithm.GameOfLife
import gameoflife.helper.*
import gameoflife.helper.SeedHelper

import javax.swing.JFrame
import javax.swing.WindowConstants
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Insets
import java.awt.KeyEventPostProcessor
import java.awt.KeyboardFocusManager
import java.awt.Toolkit
import java.awt.event.KeyEvent
import javax.swing.JPanel

class GameOfLifeUI(params: Array<String>) : JPanel(), KeyEventPostProcessor {
    var isContinue = object : ContinueCheck {
        override fun isContinue() = isContinueFlag
    }
    private val gameOfLife get() = paint.gameOfLife

    private val window = JFrame()
    private var isContinueFlag = true
    private var evolveToggle = 1
    private var automaton = true
    private var path: String? = null
    private var iteration: Int = 0
    private var waitTime: Long = 0
    private val screenSize: Dimension get() = Toolkit.getDefaultToolkit().screenSize
    private val isContinueToEvolve get() = automaton || evolveToggle == 0
    private val boundary get() = paint.boundary
    private val paint = Paint(this)

    init {
        params.takeIf { it.isNotEmpty() }
            .let { this.setup(it!!) }
    }

    private fun setup(params: Array<String>): GameOfLifeUI {
        path = params[0]
        paint.gameOfLife = buildGameOfLife(loadSeeds(path!!))
        paint.gameOfLife = gameOfLife
        automaton = params.isAutomaton
        waitTime = params.waitTime.toLong()
        return this
    }

    private fun buildGameOfLife(seeds: Array<String>): GameOfLife {
        paint.boundary = SeedHelper.getBoundaryFromHeader(seeds)
        return GameOfLife(SeedHelper.seedToMap(seeds))
    }

    fun setWaitTime(waitTime: Long): GameOfLifeUI {
        this.waitTime = waitTime
        return this
    }

    fun run() {
        setupKeyboardListener()
        setupFrame()

        while (isContinue.isContinue()) {
            repaint()
            waitAWhile()
            if (isContinueToEvolve) {
                evolve()
            }
        }
        close()
    }

    private fun close() {
        window.isVisible = false
        window.dispose()
    }

    private fun setupKeyboardListener() {
        KeyboardFocusManager.getCurrentKeyboardFocusManager().addKeyEventPostProcessor(this)
    }

    private fun setupFrame() {
        paint.cellSize = calculateCellSize

        val width = calculatePanelSize(boundary!!.x.toDouble())
        val height = calculatePanelSize(boundary!!.y.toDouble())

        setSize(width, height)
        isFocusable = true
        window.title = title(path)
        window.isResizable = false
        window.isVisible = true
        window.setBounds(width.horizontalPosition, height.verticalPosition, width.frameWidth, height.frameHeight)
        window.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        window.contentPane.add(this)
    }

    private val calculateCellSize: Int
        get() =
            calculateCellSize(screenSize.height, boundary!!.y.toDouble())
                .coerceAtMost(calculateCellSize(screenSize.width, boundary!!.x.toDouble()))
                .coerceAtMost(MAX_CELL_SIZE)
                .coerceAtLeast(MIN_CELL_SIZE)

    private fun calculateCellSize(screenSize: Int, numberOfCells: Double) =
        ((screenSize * 3).toDouble() / 4.0 / numberOfCells).toInt()

    private fun calculatePanelSize(position: Double) = (position * cellSize!!).toInt()

    private val Int.horizontalPosition get() = calculatePosition(screenSize.width, frameWidth)
    private val Int.verticalPosition get() = calculatePosition(screenSize.height, frameHeight)
    private fun calculatePosition(screenSize: Int, frameSize: Int) = (screenSize - frameSize) / 2

    private val Int.frameWidth get() = this + calculateInsertsValue { it.left + it.right }
    private val Int.frameHeight get() = this + calculateInsertsValue { it.top + it.bottom }
    private fun calculateInsertsValue(calculate: (Insets) -> Int) = calculate(window.insets)

    private fun evolve() {
        paint.tick()
        evolveToggle++
        window.title = "$path - #$iteration"
        iteration++
    }

    private val cellSize get() = paint.cellSize
    private fun title(path: String?) = "Seed: $path"

    override fun paint(graphics: Graphics) = paint.paint(graphics)

    private fun waitAWhile() {
        if (waitTime > 0) Thread.sleep(waitTime)
    }

    override fun postProcessKeyEvent(e: KeyEvent): Boolean {
        var result = false
        when (e.keyCode) {
            KeyEvent.VK_ESCAPE -> isContinueFlag = false
            KeyEvent.VK_SPACE -> {
                automaton = e.isControlDown
                evolveToggle = if (evolveToggle < 2) evolveToggle + 1 else 0
                e.consume()
                result = true
            }
        }
        return result
    }

    companion object {
        const val MAX_CELL_SIZE = 100
        private const val MIN_CELL_SIZE = 6

        @JvmStatic
        fun main(params: Array<String>) {
            GameOfLifeUI(params).run()
        }
    }
}